include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,297 @@
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_BATCH_SOLVER_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_BATCH_SOLVER_HPP_
+
+
+#include <ginkgo/core/base/batch_lin_op.hpp>
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/base/utils_helper.hpp>
+#include <ginkgo/core/log/batch_logger.hpp>
+#include <ginkgo/core/matrix/batch_identity.hpp>
+
+
+namespace gko {
+namespace batch {
+namespace solver {
+
+
+class BatchSolver {
+public:
+    /**
+     * Returns the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const BatchLinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Returns the generated preconditioner.
+     *
+     * @return the generated preconditioner.
+     */
+    std::shared_ptr<const BatchLinOp> get_preconditioner() const
+    {
+        return preconditioner_;
+    }
+
+    /**
+     * Get the residual tolerance used by the solver.
+     *
+     * @return The residual tolerance.
+     */
+    double get_residual_tolerance() const { return residual_tol_; }
+
+    /**
+     * Update the residual tolerance to be used by the solver.
+     *
+     * @param res_tol  The residual tolerance to be used for subsequent
+     *                 invocations of the solver.
+     */
+    void set_residual_tolerance(double res_tol) { residual_tol_ = res_tol; }
```

Comment:
Are these setters required?

Reply:
Yes, these were requested by applications.

Reply:
But why should they re-create the solvers if they want to change some parameters of a solver ? Solvers dont have a fixed state. I think it brings unneccessary complexity in their code base. For example, if they want to reset only the tolerance, then they will have to specify all the parameters again.

Reply:
Is `solver_a->get_parameters().with_one_parameter_changed(...).on(exec)->generate(M);` valid code ? `solver_a` is an object of type solver and not a factory anymore. I think the users will have to pass the factory and for cases where regeneration is not cheap, this would be significantly expensive.

Reply:
You're right `get_parameters` returns a const reference, so users would need to copy it. Or we make it return a copy directly in a later release. In any case, for now we can't change it, since users already rely on it. But IMO we should aim to change this in a future release.